lib/elixir/lib/file.ex

```diff
@@ -1340,6 +1355,19 @@ defmodule File do
   defp normalize_modes([], true), do: [:binary]
   defp normalize_modes([], false), do: []
 
+  defp normalize_write_modes(modes) do
+    normalize_modes([:write] ++ modes, true)
+    |> set_raw_unless_encoding_specified
+  end
+
+  # Using raw mode means we get the performance benefit of using `writev`.
+  # However, Erlang will raise a badarg error if a file encoding is specified
+  # and raw mode is also requested.
+  # http://erlang.org/doc/man/file.html#open-2
+  defp set_raw_unless_encoding_specified(modes) do
+    if Keyword.has_key?(modes, :encoding), do: modes, else: [:raw | modes]
```

Comment:
Good call. If we all we are going to do is to avoid make_binary on raw
mode, then I think we should send this patch directly to OTP indeed.

On Wednesday, August 24, 2016, James Fish <notifications@github.com> wrote:

> In lib/elixir/lib/file.ex
> <https://github.com/elixir-lang/elixir/pull/5154#discussion_r76100140>:
>
> > @@ -1340,6 +1355,19 @@ defmodule File do
> >    defp normalize_modes([], true), do: [:binary]
> >    defp normalize_modes([], false), do: []
> >
> > +  defp normalize_write_modes(modes) do
> > +    normalize_modes([:write] ++ modes, true)
> > +    |> set_raw_unless_encoding_specified
> > +  end
> > +
> > +  # Using raw mode means we get the performance benefit of using `writev`.
> > +  # However, Erlang will raise a badarg error if a file encoding is specified
> > +  # and raw mode is also requested.
> > +  # http://erlang.org/doc/man/file.html#open-2
> > +  defp set_raw_unless_encoding_specified(modes) do
> > +    if Keyword.has_key?(modes, :encoding), do: modes, else: [:raw | modes]
>
> On a diskless node connected to a master node where the file server
> process is.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/elixir-lang/elixir/pull/5154/files/332a6abe10eddce03064dd5f21e61c3e9a97a17c#r76100140>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAlbgG1nYUzgciLWimib6gP2caOfliKks5qjH1VgaJpZM4JsE6I>
> .
>


-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*